### PR TITLE
feat: Implementation of the module

### PR DESCRIPTION
**Pull request by rubenVB01**
## Description

This PR contains the implementation of the push-based-log-export module. Specifically, the main.tf, variables.tf, versions.tf and outputs.tf. 
For the case of variables.tf and outputs.tf, also contains the descriptions of inputs and outputs.

### CODE REVIEW

File: modules/s3_bucket/main.tf [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+  # need to parse the date because bucket names do not accept ":"
+  bucket_name_def          = var.bucket_name == null ? "${var.project_id}-atlas-push-log-${formatdate("YY-M-D-h-m-s", timestamp())}" : var.bucket_name
+  iam_role_name_def        = var.iam_role_name == null ? "${var.project_id}-push-based-log-export-role" : var.iam_role_name
+  iam_role_policy_name_def = var.iam_role_policy_name == null ? "${var.project_id}-push-based-log-export-policy" : var.iam_role_policy_name
+}
+
+resource "aws_s3_bucket" "s3_bucket" {
+  count         = var.create_bucket ? 1 : 0
+  bucket        = local.bucket_name_def
+  force_destroy = true
```
Comment (rubenVB01) [inline modules/s3_bucket/main.tf]: This still has to be discussed (whether force_destroy should be true or false)

File: modules/s3_bucket/outputs.tf [hunk 2]
```diff
@@ -0,0 +1,37 @@
+# mongodbatlas_cloud_provider_access_setup
+
+output "aws_configuration" {
+  description = "AWS related ARN roles. Contains: atlas_assumed_role_external_id, unique external ID Atlas uses when assuming the IAM role in your AWS account and atlas_aws_account_arn, ARN associated with the Atlas AWS account used to assume IAM roles in your AWS account."
```
Comment (corryroot) [inline modules/s3_bucket/outputs.tf]:
```suggestion
  description = "AWS related ARN role configuration that contains the atlas_assumed_role_external_id and atlas_aws_account_arn. atlas_assumed_role_external_id describes the unique external ID that Atlas uses when it assumes the IAM role in your AWS account. atlas_aws_account_arn  describes the ARN associated with the Atlas AWS account used to assume IAM roles in your AWS account."
```

File: modules/s3_bucket/outputs.tf [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+  value       = mongodbatlas_cloud_provider_access_setup.setup.aws_config
+}
+
+# aws_iam_role
+
+output "role_arn" {
+  description = "Amazon Resource Name (ARN) specifying the role."
```
Comment (corryroot) [inline modules/s3_bucket/outputs.tf]:
```suggestion
  description = "Amazon Resource Name (ARN) that specifies the role."
```

File: modules/s3_bucket/outputs.tf [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+  value       = aws_iam_role.iam_role.arn
+}
+
+output "unique_id" {
+  description = "Stable and unique string identifying the role."
```
Comment (corryroot) [inline modules/s3_bucket/outputs.tf]:
```suggestion
  description = "Stable and unique string that identifies the role."
```

File: modules/s3_bucket/outputs.tf [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+  value       = aws_iam_role.iam_role.unique_id
+}
+
+# aws_iam_role_policy
+
+output "role_policy_id" {
+  description = "The role policy ID, in the form of role_name:role_policy_name."
```
Comment (corryroot) [inline modules/s3_bucket/outputs.tf]:
```suggestion
  description = "Role policy ID, in the form of role_name:role_policy_name."
```

File: modules/s3_bucket/outputs.tf [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+
+# aws_s3_bucket (data source)
+
+output "bucket_name" {
+  description = "Name of the bucket."
+  value       = data.aws_s3_bucket.s3_bucket_data.id
+}
+
+output "bucket_arn" {
+  description = "ARN of the bucket. Will be of format arn:aws:s3:::bucketname."
```
Comment (corryroot) [inline modules/s3_bucket/outputs.tf]:
```suggestion
  description = "ARN of the bucket, in the form of arn:aws:s3:::bucketname."
```

File: modules/s3_bucket/variables.tf [hunk 7]
```diff
@@ -0,0 +1,31 @@
+variable "project_id" {
+  description = "The project id (e.g., 65def6ce0f722a1507105bb5)."
+  type        = string
+}
+
+variable "bucket_name" {
+  description = "The name of the bucket to which Atlas will send the logs to. If no value is provided, a default value of the form {project_id}-atlas-push-log-{timestamp} is assigned (e.g., 66a7ae173df3c34412a71cd6-atlas-push-log-24-7-29-15-0-38)."
```
Comment (corryroot) [inline modules/s3_bucket/variables.tf]:
```suggestion
  description = "Name of the bucket that Atlas sends the logs to. If you provide no value , a default value of the form {project_id}-atlas-push-log-{timestamp} is assigned (e.g., 66a7ae173df3c34412a71cd6-atlas-push-log-24-7-29-15-0-38)."
```

File: modules/s3_bucket/variables.tf [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+}
+
+variable "create_bucket" {
+  description = "Flag that determines whether the bucket resource has to be created or not."
+  type        = bool
+  default     = true
+}
+
+variable "iam_role_name" {
+  description = "The name of the IAM role to use to set up cloud provider access in Atlas. If no value is provided, a default value of the form {project_id}-push-based-log-export-role is assigned (e.g., 66a7ae173df3c34412a71cd6-push-based-log-export-role)."
```
Comment (corryroot) [inline modules/s3_bucket/variables.tf]:
```suggestion
  description = "Name of the IAM role to use to set up cloud provider access in Atlas. If you provide no value, a default value of the form {project_id}-push-based-log-export-role is assigned (e.g., 66a7ae173df3c34412a71cd6-push-based-log-export-role)."
```

File: modules/s3_bucket/variables.tf [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+  type        = string
+  default     = null
+  nullable    = true
+}
+
+variable "iam_role_policy_name" {
+  description = "The name of the IAM role policy for the configured aws_iam_role_name. If no value is provided, a default value of the form {project_id}-push-based-log-export-policy is assigned (e.g., 66a7ae173df3c34412a71cd6-push-based-log-export-policy)."
```
Comment (corryroot) [inline modules/s3_bucket/variables.tf]:
```suggestion
  description = "Name of the IAM role policy for the configured aws_iam_role_name. If you provide no value, a default value of the form {project_id}-push-based-log-export-policy is assigned (e.g., 66a7ae173df3c34412a71cd6-push-based-log-export-policy)."
```

**Review by corryroot — APPROVED** [review]
LGTM! I commented with some copy nits.

File: modules/s3_bucket/main.tf [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+  force_destroy = var.force_destroy
+}
+
+data "aws_s3_bucket" "s3_bucket_data" {
+  bucket = var.create_bucket ? aws_s3_bucket.s3_bucket[0].id : local.bucket_name_def
+}
+
+resource "aws_iam_role" "iam_role" {
+  name                 = local.iam_role_name_def
+  max_session_duration = 43200
```
Comment (AgustinBettati) [inline modules/s3_bucket/main.tf:20]: Curious about this value, is it a default? Does this imply any regeneration of the session from user side?

File: modules/s3_bucket/versions.tf [hunk 11]
```diff
@@ -0,0 +1,11 @@
+terraform {
+  required_providers {
+    mongodbatlas = {
+      source = "mongodb/mongodbatlas"
```
Comment (AgustinBettati) [inline modules/s3_bucket/versions.tf:4]: would make sure to put a constraint of higher or equal to 1.16.0 ([reference](https://github.com/mongodb/terraform-provider-mongodbatlas/blob/master/CHANGELOG.md#1160-april-29-2024))

File: modules/s3_bucket/main.tf (hunk 10 above)
Comment (rubenVB01) [reply modules/s3_bucket/main.tf:20]: When I first tried the module without specifying this value I got an error regarding exceeding the session duration, so I got the value from the [example](https://github.com/mongodb/terraform-provider-mongodbatlas/blob/master/examples/mongodbatlas_push_based_log_export/aws.tf) in the provider. 
This is the description I found about the attribute: _Maximum session duration (in seconds) that you want to set for the specified role. If you do not specify a value for this setting, the default maximum of one hour is applied. This setting can have a value from 1 hour to 12 hours._

Comment (AgustinBettati) [reply modules/s3_bucket/main.tf:20]: @maastha do you have any additional context of this value? Maybe we can go with the default value

Comment (maastha) [reply modules/s3_bucket/main.tf:20]: Don't remember much context, but I got this from MMS, I think here https://github.com/10gen/mms/blob/5b6f0a71de606520d390ed7adab09db64d8bac06/client/packages/project/nds/security/components/EnterpriseSecurity/PushBasedLogging/PushBasedLoggingModal.tsx#L64 

@rubenVB01 could you share the error you get if we don't specify this? Does the error come from push_based_log_export resource or from cloud access setup? Based on that we can look into reaching out to API team to know more about the defaults

Comment (rubenVB01) [reply modules/s3_bucket/main.tf:20]: Yes, I attach the error that it makes. Apparently is from the push-based-log-export resource. I just noticed that there's an optional attribute in the resource, which is [timeouts](https://registry.terraform.io/providers/mongodb/mongodbatlas/latest/docs/resources/push_based_log_export#nested-schema-for-timeouts), however I don't see which are the default values.
Screenshot 2024-07-31 at 14 23 11

File: modules/s3_bucket/main.tf (hunk 1 above)
Comment (rubenVB01) [reply modules/s3_bucket/main.tf]: In the end it's an input variable with default value false

**Review by AgustinBettati — APPROVED** [review]
LGTM, we can follow up with the Atlas team supporting push based logging to get more insight on the max session duration needed in the iam role

**Comment by rubenVB01** [comment]
I'm going to merge, can do follow up PRs if there are additional reviews

File: modules/s3_bucket/main.tf (hunk 10 above)
Comment (rubenVB01) [reply modules/s3_bucket/main.tf:20]: _by default PBLE export requires a max role duration of 12 hours._.  This explains the 43200 value (12 hrs in secs)

**Pull request merged**